Feat/pivotal api

kuderluka opened the pull request.

kuderluka committed:
Added needed services

kuderluka committed:
Added service that handles api calls

kuderluka committed:
Added controller

kuderluka committed:
Update .env.example

kuderluka committed:
Added custom command

kuderluka committed:
Added additional const

kuderluka committed:
Added custom command

kuderluka committed:
Added comments

vojkovoga commented on newsletter_laravel/.env.example:
```diff
@@ -4,6 +4,10 @@ APP_KEY=
 APP_DEBUG=true
 APP_URL=http://localhost
 
+PIVOTAL_API_TOKEN=627d82676e1f0a342fb974949da5599c
```
In `.env.example` you don't set values. Especially not correct ones. You just define **variables**.

vojkovoga commented on newsletter_laravel/app/Http/Controllers/NewsletterController.php:
```diff
@@ -0,0 +1,10 @@
+
+use Illuminate\Http\Request;
+
+class NewsletterController extends Controller
```
Empty `Controller` ?

vojkovoga commented on newsletter_laravel/app/Console/Commands/ExportNewsletterStories.php:
```diff
@@ -0,0 +1,59 @@
+use Symfony\Component\Console\Style\SymfonyStyle;
+use League\Csv\Writer;
+
+class ExportNewsletterStories extends Command
```
Rename this to append `Command` name to `ExportNewsletterStoriesCommand`

vojkovoga commented on newsletter_laravel/app/Services/NewsletterService.php:
```diff
@@ -0,0 +1,102 @@
+use Exception;
+use Symfony\Component\Console\Style\SymfonyStyle;
+
+class NewsletterService {
```
Better naming would be `PivotalTrackerClient` since you are currently selecting only `Newsletter` stories, but it's possible you will need do additional requests.

vojkovoga replied:
Missing small comment for this class.

vojkovoga commented on newsletter_laravel/app/Services/NewsletterService.php:
```diff
@@ -0,0 +1,102 @@
+
+class NewsletterService {
+
+    private Client $client;
```
Missing comment

vojkovoga commented on newsletter_laravel/app/Services/NewsletterService.php:
```diff
@@ -0,0 +1,102 @@
+
+    private SymfonyStyle $io;
+
+    public function __construct()
```
Missing comment

vojkovoga commented on newsletter_laravel/app/Services/NewsletterService.php:
```diff
@@ -0,0 +1,102 @@
+
+    private Client $client;
+
+    private SymfonyStyle $io;
```
You don't need this here, since this class **MUST NOT** output anything to console. If there is an error, it should throw an `Exception` and you should handle that in different place.

vojkovoga commented on newsletter_laravel/app/Services/NewsletterService.php:
```diff
@@ -0,0 +1,102 @@
+     * @return array|void|null
+     * @throws \GuzzleHttp\Exception\GuzzleException
+     */
+    public function exportStories(SymfonyStyle $io)
```
Remove parameter and add return type

vojkovoga commented on newsletter_laravel/app/Services/NewsletterService.php:
```diff
@@ -0,0 +1,102 @@
+     * @param $stories
+     * @return array
+     */
+    public function extractForCSV($stories)
```
Add type to parameter and add return type

vojkovoga commented on newsletter_laravel/app/Services/NewsletterService.php:
```diff
@@ -0,0 +1,102 @@
+                'labels' => $labelsString,
+            ];
+        }
+        return $output;
```
Add empty space above this line.

vojkovoga commented on newsletter_laravel/app/Services/NewsletterService.php:
```diff
@@ -0,0 +1,102 @@
+    {
+        $this->io = $io;
+        try {
+            $response = $this->client->request('GET', '/services/v5/projects/' . env('PIVOTAL_PROJECT_ID') . '/stories', [
```
Extract this method call to new method like `loadStoriesForProject(int $projectId): array`

vojkovoga commented on newsletter_laravel/app/Services/NewsletterService.php:
```diff
@@ -0,0 +1,102 @@
+        try {
+            $output = [];
+            foreach($data as $story) {
+                $response = $this->client->request('GET', '/services/v5/projects/' . env('PIVOTAL_PROJECT_ID') . '/stories/' . $story['id'] . '/reviews', [
```
Extract this method to something like `loadReviewsForStory(int $projectId, int $storyId): array`

vojkovoga commented on newsletter_laravel/app/Services/NewsletterService.php:
```diff
@@ -0,0 +1,102 @@
+     */
+    public function exportStories(SymfonyStyle $io)
+    {
+        $this->io = $io;
```
This is not needed

vojkovoga commented on newsletter_laravel/app/Services/NewsletterService.php:
```diff
@@ -0,0 +1,102 @@
+                ]
+            ]);
+
+            $data = $this->filterByReviews(json_decode($response->getBody(), true));
```
You don't need extra variable here, just `return $this->filter...`

vojkovoga commented on newsletter_laravel/app/Services/NewsletterService.php:
```diff
@@ -0,0 +1,102 @@
+            $data = $this->filterByReviews(json_decode($response->getBody(), true));
+            return $data;
+
+        } catch (RequestException $e) {
```
You don't need `try-catch` statement here.

vojkovoga commented on newsletter_laravel/app/Services/NewsletterService.php:
```diff
@@ -0,0 +1,102 @@
+            }
+            return $output;
+
+        } catch (Exception $e) {
```
You don't need `try-catch` statement here.

vojkovoga commented on newsletter_laravel/app/Console/Commands/ExportNewsletterStories.php:
```diff
@@ -0,0 +1,59 @@
+     */
+    protected $description = 'Export all the stories with Newsletter review type';
+
+    public function __construct(NewsletterService $newsletterService)
```
This parameter can be moved to `handle` method like `handle(NewsletterService $...)`. So you can completely remove constructor here.

vojkovoga commented on newsletter_laravel/app/Console/Commands/ExportNewsletterStories.php:
```diff
@@ -0,0 +1,59 @@
+     */
+    public function handle()
+    {
+        $io = new SymfonyStyle($this->input, $this->output);
```
You don't need this

vojkovoga commented on newsletter_laravel/app/Console/Commands/ExportNewsletterStories.php:
```diff
@@ -0,0 +1,59 @@
+        $io = new SymfonyStyle($this->input, $this->output);
+
+        try {
+            $data = $this->newsletterService->exportStories($io);
```
Instead of `data`, write something that makes sense like `stories`

vojkovoga commented on newsletter_laravel/app/Console/Commands/ExportNewsletterStories.php:
```diff
@@ -0,0 +1,59 @@
+            $data = $this->newsletterService->exportStories($io);
+            $csvData = $this->newsletterService->extractForCSV($data);
+
+            $csvFilename = 'newsletter_stories';
```
Writing to `CSV` file extract to different method

vojkovoga commented on newsletter_laravel/app/Console/Commands/ExportNewsletterStories.php:
```diff
@@ -0,0 +1,59 @@
+                $csv->insertOne($rowData);
+            }
+
+            $io->success('Data exported to CSV: ' . $csvPath);
```
Laravel `Command` class has access to `$this->success` or `$this->error`, so you don't need to instantiate new `SymfonyStyle` first. Check Laravel's `InteractsWithIO.php` file

vojkovoga commented on newsletter_laravel/app/Services/NewsletterService.php:
```diff
@@ -0,0 +1,102 @@
+        try {
+            $response = $this->client->request('GET', '/services/v5/projects/' . env('PIVOTAL_PROJECT_ID') . '/stories', [
+                'headers' => [
+                    'X-TrackerToken' => env('PIVOTAL_API_TOKEN'),
```
It's bad practice to use here `env` method. Implement config file like `config/pivotal-tracker.php` and then call here `config('pivotal-tracker.api_token')`

kuderluka replied:
This was done like that due to being required in the instructions: 
!! API tokens and usernames must be saved in .env file

vojkovoga commented on newsletter_laravel/app/Services/NewsletterService.php:
```diff
@@ -0,0 +1,102 @@
+            foreach($data as $story) {
+                $response = $this->client->request('GET', '/services/v5/projects/' . env('PIVOTAL_PROJECT_ID') . '/stories/' . $story['id'] . '/reviews', [
+                    'headers' => [
+                        'X-TrackerToken' => env('PIVOTAL_API_TOKEN'),
```
It's bad practice to use here `env` method. Implement config file like `config/pivotal-tracker.php` and then call here `config('pivotal-tracker.api_token')`

kuderluka committed:
Removed values

kuderluka committed:
Unnecessary controller deleted

kuderluka committed:
Added config file

kuderluka committed:
File renamed

Besides renaming various bugfixes. syntax errors and fragmentation were performed

kuderluka committed:
Added comments

kuderluka committed:
env() calls in config file

kuderluka committed:
Syntax
